sdk/internal/perf/perf_runner.go

```diff
@@ -0,0 +1,432 @@
+	opsPerSecond := r.opsPerSecond(warmup)
+	if opsPerSecond == 0.0 {
+		return fmt.Errorf("completed without generating operation statistics")
+	}
+
+	secondsPerOp := 1.0 / opsPerSecond
+	weightedAvg := float64(totalOperations) / opsPerSecond
+
+	if warmup {
+		fmt.Println("\n=== Warm Up Results ===")
+	} else {
+		fmt.Println("\n=== Results ===")
+	}
+	fmt.Printf(
+		"Completed %s operations in a weighted-average of %ss (%s ops/s, %s s/op)\n",
+		r.messagePrinter.Sprintf("%d", totalOperations),
+		r.messagePrinter.Sprintf("%.2f", weightedAvg),
+		r.messagePrinter.Sprintf("%.3f", opsPerSecond),
+		r.messagePrinter.Sprintf("%.3f", secondsPerOp),
+	)
+	return nil
+}
+
+// runTest takes care of the semantics of running a single iteration.
+// It changes configuration on the proxy, increments counters, and
+// updates the running-time.
+func (r *perfRunner) runTest(p PerfTest, index int, ID string) {
+	defer wg.Done()
+	if debug {
+		log.Printf("number of proxies %d", len(r.proxyTransports))
+	}
+
+	r.allOptions.mu.Lock()
+	opts := r.allOptions.opts[index]
+	r.allOptions.mu.Unlock()
+
+	// If we are using the test proxy need to set up the in-memory recording.
+	if testProxyURLs != "" {
+		// First request goes through in Live mode
+		r.proxyTransports[ID].SetMode("live")
+		err := p.Run(context.Background())
+		if err != nil {
+			if err != nil {
+				panic(err)
+			}
+		}
+
+		// 2nd request goes through in Record mode
+		r.proxyTransports[ID].SetMode("record")
+		err = r.proxyTransports[ID].start()
+		if err != nil {
+			panic(err)
+
+		}
+
+		err = p.Run(context.Background())
+		if err != nil {
+			if err != nil {
+				panic(err)
+			}
+		}
+		err = r.proxyTransports[ID].stop()
+		if err != nil {
+			panic(err)
+		}
+
+		// All ensuing requests go through in Playback mode
+		r.proxyTransports[ID].SetMode("playback")
+		err = r.proxyTransports[ID].start()
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	if warmUpDuration > 0 {
+		opts.warmupStart = time.Now()
+
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*time.Duration(warmUpDuration))
+		defer cancel()
+
+		lastSavedTime := time.Now()
+		for time.Since(opts.warmupStart).Seconds() < float64(warmUpDuration) {
+			err := p.Run(ctx)
+			if err != nil {
+				if errors.Is(err, context.DeadlineExceeded) {
+					break
+				} else {
+					panic(err)
+				}
+			}
+			opts.incrememt(true)
+
+			if time.Since(lastSavedTime).Seconds() > 0.3 {
+				opts.warmupElapsed.SetFloat(time.Since(opts.warmupStart).Seconds())
+				lastSavedTime = time.Now()
+			}
+		}
+
+		opts.warmupElapsed.SetFloat(time.Since(opts.warmupStart).Seconds())
+	}
+	val := atomic.AddInt32(&r.warmupFinished, 1)
+	if debug {
+		fmt.Printf("finished %d warmups\n", val)
+	}
+
+	opts.runStart = time.Now()
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*time.Duration(duration))
+	defer cancel()
+
+	lastSavedTime := time.Now()
+	for time.Since(opts.runStart).Seconds() < float64(duration) {
```

Comment:
Just briefly eyeballing this but it looks like warmup and this block are basically identical, except you'd probably parameterize which 'opts.<elapsed>' you'd update. Probably worth consolidating a bit. 

Might also give you a simple testable function afterwards, as well.